docs/rfcs/0001_modularize_iceberg_implementations.md

```diff
@@ -0,0 +1,196 @@
+
+After recent maintainer discussions we agreed on two principles:
+1. The `iceberg` crate itself remains the single source of truth for all protocol traits and data structures.
+2. All concrete integrations (Tokio runtime, opendal `FileIO`, Arrow/DataFusion executors, catalog adapters, etc.) move out of `iceberg` into dedicated companion crates. Users who need a ready-made execution path can depend on those crates (for example `iceberg-datafusion`) while users building custom stacks can depend solely on `iceberg`.
+
+This RFC describes the plan to slim down `iceberg` into a pure protocol crate and to reorganize the workspace around pluggable companion crates.
+
+## Goals and Scope
+
+- **Keep `iceberg` as the protocol crate**: it exposes all traits (`Catalog`, `Table`, `Transaction`, `FileIO`, `Runtime`, `ScanPlan`, etc.) plus metadata/plan logic, but no longer ships concrete runtimes or storage adapters.
+- **Detach embedded implementations**: move opendal-based IO, Tokio runtime helpers, Arrow converters, and similar code into separate crates under `crates/fileio/*`, `crates/runtime/*`, `crates/engine/*`, or existing integration crates.
+- **Enable composable combinations**: users assemble the stack they need by combining `iceberg` with specific implementation crates (e.g., `iceberg-fileio-opendal`, `iceberg-runtime-tokio`, `iceberg-engine-arrow`, `iceberg-datafusion`).
+- **Minimize breaking surfaces**: trait APIs stay in `iceberg`; downstream crates only adjust their dependency graph.
+
+Out of scope: changing the Iceberg table specification or rewriting catalog adapters’ external behavior.
+
+## Architecture Overview
+
+### Workspace Layout
+
+```
+crates/
+  iceberg/                # core traits, metadata, planning, transactions
+  fileio/
+    opendal/             # e.g. `iceberg-fileio-opendal`
+    fs/                  # other FileIO implementations
+  runtime/
+    tokio/               # e.g. `iceberg-runtime-tokio`
+    smol/
+  engine/
+    arrow/               # Arrow executor & schema helpers
+  catalog/*              # catalog adapters (REST, HMS, Glue, etc.)
+  integrations/
+    datafusion/          # combines core + implementations for DF
+    cache-moka/
+    playground/
+```
+
+- `crates/iceberg` no longer depends on opendal, Tokio, Arrow, or DataFusion.
+- Implementation crates depend on `iceberg` to get the trait surfaces they implement.
+- Higher-level crates (e.g., `iceberg-datafusion`) pull in the required runtime/FileIO/executor crates and expose an opinionated combination.
+
+### Core Trait Surfaces (within `iceberg`)
+
+#### FileIO
```

Comment:
As we said before, this may not be a good place to discuss details about `FileIO`.  We should have another rfc to discuss this.